spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/AnalyzeTableSparkAction.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.spark.actions;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.iceberg.StatisticsFile;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.actions.AnalyzeTable;
+import org.apache.iceberg.actions.ImmutableAnalyzeTable;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.puffin.StandardBlobTypes;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.spark.JobGroupInfo;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+import org.apache.spark.sql.SparkSession;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Computes the statistic of the given columns and stores it as Puffin files. */
```

Comment:
`AnalyzeTableSparkAction` is a generic name, I see that in future we want to compute the partition stats too. Which may not be written as puffin files. 

Either we can change the change the naming to `computeNDVSketches` or make it generic such that any kind of stats can be computed from this.

Reply:
I tried to follow the model of RDMS and Engines like Trino using `ANALYZE TABLE <tblName>` to collect all table level stats.
With a  procedure per stats model, the user have to invoke procedure/action for every stats  and 
also with any new stats addition, the user need to ensure to update his code to call the new procedure/action.

> not mix it with partition stats.

I think we could have partition stats as a separate action since it per partition, whereas this procedure can collect top level table stats.

Reply:
@karuppayya 
I can see the tests in TestAnalyzeTableAction, it's working fine.
But have we tested in Spark, whether its working with a query like -
"Analyze table table1 compute statistics" ?

Because generally it gives the error
"[NOT_SUPPORTED_COMMAND_FOR_V2_TABLE] ANALYZE TABLE is not supported for v2 tables."

Reply:
Spark doesnot have the grammar for Analyzing tables.
This PR introduces a Spark action. In subsequent PR, i plan to introduce a iceberg procedure to invoke the Spark action.

Reply:
Actually I think Spark has a grammar, and be great to plug it into there.  https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala#L428

Reply:
> I see that in future we want to compute the partition stats too. Which may not be written as puffin files.

Hi, @ajantha-bhat I agree with you, otherwise, the queries would have a lot of limitations, such as being applicable only for calculating the NDV over the entire table.

For example, Trino might want to read the NDV values written by Spark to respond to queries. However, if the query has partition filter conditions, then Trino would not be able to use the pre-computed NDV information from Spark.

Reply:
Hm this discussion makes me wonder if we're under spec'd in this regard. 

https://iceberg.apache.org/puffin-spec/#blob-types

```
The blob metadata for this blob may include following properties:
    ndv: estimate of number of distinct values, derived from the sketch.
```

It really seems like we should take a stance. Either it must be in the sketch or it must be in the properties. "may include" seems a little too loose.

Reply:
> Do we need to add the property `ndv` , should nt engines be reading the value from the sketch?

yes, this property should be declared and therefore present to the query engine without opening the stats file
this is why blob properties exists in the first place

i thinking that the wording for properties could be changed from "may" to "should"
https://iceberg.apache.org/puffin-spec/#apache-datasketches-theta-v1-blob-type

if we don't set the property, Trino won't use the NDV values
https://github.com/trinodb/trino/blob/a69fa9f2ae5370217339cc7223bb3d0db97470da/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsReader.java#L255

Reply:
> For example, Trino might want to read the NDV values written by Spark to respond to queries. However, if the query has partition filter conditions, then Trino would not be able to use the pre-computed NDV information from Spark. 

@rice668 Trino will use table label NDV information regardless of potential partitioning filters ([code](https://github.com/trinodb/trino/blob/a69fa9f2ae5370217339cc7223bb3d0db97470da/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsReader.java#L255)).

we will revisit when partition stats feature (https://github.com/apache/iceberg/issues/8450) matures

Reply:
> i thinking that the wording for properties could be changed from "may" to "should"

+1, even though the spec called this out as `not` in required prop, would be nice if we can make this `required` 
- I do know some more internal readers (we got inspired by trino, also at that time trino was the only known writer for these files) that take dependency on the property as parsing the datasketch per query would be expensive or would have to maintain it at a cache for it per execution.
- It would not be too much of a work for writers to calculate this in this first place.

Reply:
Great, seems like there's consensus on making the property required. I also agree with that since practically it makes sense for adoption to not have to force different engines to have to open up the sketch which as @singhpk234 mentioned adds complexity.

I'll raise a PR for the spec update if there's no objections?

Reply:
> I'll raise a PR for the spec update if there's no objections?

thanks!

Reply:
@karuppayya Thanks for the great work! Sorry I didn't have time to take a look at your PR earlier. For ANALYZE table, Spark has the following syntax:
```
ANALYZE TABLE table_identifier [ partition_spec ]
    COMPUTE STATISTICS [ NOSCAN | FOR COLUMNS col [ , ... ] | FOR ALL COLUMNS ]
```
For column-level stats, Spark computes 

- NDV
- Max (for numeric, Date, and Timestamp only)
- Min  (for numeric, Date, and Timestamp only)
- Null Count
- Avg Len
- Max Len

We probably want to ensure that Iceberg implementation aligns with Spark's current functionality.

Reply:
Hi @huaxingao do you mean, that this PR should take care of the other stats too apart from NDV, like MIN, MAX, NULL Count, etc ?

Reply:
Hi @karuppayya are we planning to induce the other column-level stats in this PR

Reply:
@jeesou , This PR is only for the NDV stats. ([PR](https://github.com/apache/iceberg/pull/10659) to propagate the stats in scan)

Reply:
@karuppayya @huaxingao following up on this discussion and issue #10791.
Wouldn't it make sense to extend the current method which extracts the theta sketches to also extract the rest of the statistics by enhancing the select expression [here](https://github.com/apache/iceberg/blob/7172760101984045245a3e759e601a060dbcc3b5/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/NDVSketchGenerator.java#L99).
My thinking is that instead of having a custom code for generating the min/max we would rely on spark pushing it down to iceberg. Also since we already pay the cost of generating the NDV stats it doesn't seem to like the overhead will be much.

Any thoughts? This can be still done in a separate PR that will follow this one.

Reply:
@karuppayya I would appreciate your input on the above.
If that make sense we can create a separate PR for this.

Reply:
@guykhazma , here is the [related thread](https://lists.apache.org/thread/xfpqg98z00dmlglmlhzj69q6lzcpk2xz) discussing about some aspects of storing min/max/null counts.